modules/axp/Animal/lib/Animal.pm

```diff
@@ -22,76 +26,55 @@ our $VERSION = '0.01';
 
 Don't use this class directly; instead instantiate subclasses of it.
 
-=head1 EXPORT
-
 =head1 SUBROUTINES/METHODS
 
-=head2 named
-
-Construct a new Animal, with the given name
-
-=cut
-
-sub named {
-	ref ( my $class = shift ) and croak "Static constructor used as instance call";
-	my $name = shift or croak "Need to provide a name";
-	my $self = { Name => $name, Colour => $class->default_colour };
-	return bless $self, $class;
-}
-
 =head2 name
 
 Get or set the Animal's name
 
 =cut
 
-sub name {
-	ref ( my $self = shift ) or croak "Instance variable needed";
-	if ( @_ ) {
-		$self->{ Name } = shift;
-		return $self;
-	}
-	else {
-		return $self->{ Name };
-	}
-}
+has 'name' => ( is => 'rw', required => 1 );
+
+before 'name' => $LivingCreature::__instance_check;
```

Comment:
These little hooks replace the `ref ( my $self = shift ) or croak` first line of every instance method. Handy!